src/main/java/org/truffleruby/core/proc/ProcNodes.java

```diff
@@ -175,9 +176,27 @@ protected RubyProc dup(RubyProc proc,
 
         @Specialization
         protected int arity(RubyProc proc) {
-            return proc.sharedMethodInfo.getArity().getArityNumber();
-        }
+            Arity arity = proc.sharedMethodInfo.getArity();
+            if (proc.type == ProcType.PROC) {
+                // Based on copy-past from Arity#computeArityNumber():
+                int count = arity.getRequired();
+
+                if (arity.acceptsKeywords() && !arity.allKeywordsOptional()) {
+                    count++;
+                }
+
+                // This is only one line we have to change:
+                // Was: if (optional > 0 || hasRest || (acceptsKeywords() && allKeywordsOptional)) {
+                if (arity.hasRest()) {
+                    count = -count - 1;
+                }
 
+                return count;
```

Comment:
How about storing `private final int procArityNumber;` in Arity, and compute it in the constructor with a call to `computeArityNumber(proc=true)`?
Then here we just `return proc.type == ProcType.PROC ? arity.getProcArityNumber() : arity.getArityNumber();`.

Maybe we can do better by computing the correct arity when parsing, and then changing it on proc->lambda conversion, but that sounds more complicated and not necessarily worth it.

Reply:
Yeah, I was thinking about something like this. Another one option (if we want not to store extra int value):
```java
protected int arity(RubyProc proc) {
     Arity arity = proc.sharedMethodInfo.getArity();
     if (proc.type == ProcType.PROC) {
         int count = arity.getRequired();
         if (arity.getOptional() > 0 (arity.acceptsKeywords() && arity.allKeywordsOptional)) {
             count = -count - 1; // rollback lambda effect
         } 
         return count;
     } else {
         // For lambda returns pre-build number:
         return arity.getArityNumber();
     }
}
```

Both options have their own trade-offs and clumsy enough to be honest. But I believe that if we add strong explanations (aka comments) to the code, it'll help the next developers understand the solution.

Reply:
I think it's OK to use an extra `int` field here.
That alternative is really smart but I'm concerned also harder to maintain & understand.

I might store the `Arity` in the `RubyProc` directly when reworking `RubyProc` creation (and change it when converting a proc to a lambda), but until then I think storing as an extra field is straightforward.

Reply:
@eregon correct me if I'm wrong but I just realized with these changes, we'd fix only the public arity value. The internal usage will still be broken (returns Arity number calculated for lambda). 
Therefore, I applied our idea of calculating two values to all cases where we use arity number in the code.

Reply:
That seems correct, however it would be really good to add a spec for that.
I'll add one for Hash#each_pair.

The trade-off is there are 2 fast-path usages of getArityNumber(), for Hash#each_pair and Hash#map. Those might be a bit slower if the block is not constant and is not created in the same compilation unit.

Not that the arity number is pretty much an arbitrary number and actually checking if there are enough arguments is done differently in `CheckArityNode`.

Reply:
Actually it seems impossible to test because non-lambda Procs automatically splat the 2-elements Array into multiple arguments.

Reply:
Reasoning in c5dc421, i.e., it doesn't matter.
I did find some behavior difference with an edge case, but I'd argue it's a CRuby bug because losing an argument seems pretty wrong:
```
$ ruby -ve '{a: 1}.each_pair(&proc {|a, kw: :default| p [a,kw] })'   
ruby 2.6.6p146 (2020-03-31 revision 67876) [x86_64-linux]
[:a, :default]

$ ruby -ve '{a: 1}.each_pair(&proc {|a, kw: :default| p [a,kw] })'   
truffleruby 20.3.0-dev-574d6bd2, like ruby 2.6.6, GraalVM CE Native [x86_64-linux]
[[:a, 1], :default]
```
and actually we use one arity or the other does not change behavior here.

Reply:
Filed as https://bugs.ruby-lang.org/issues/17178, seems a general issue of Proc, regardless of Hash#each_pair.

Reply:
I think this is one of those cases when you have to "repeat" the bug of CRuby. As a developer, I would not expect this behavior from `proc + array as argument`, but I think this is a significant breaking change.

Anyway, this PR (and arity itself) does not affect `proc + array as argument` result, right? You just found it by accident?

Reply:
Yeah, this PR doesn't affect that.
I am not sure we need to be bug-compatible here.
I think nothing relies on it, because that behavior seems clearly a bug and it's not been fixed yet in CRuby.

Reply:
Jeremy does not think so (see the comments to your bug report) :)